kafka.go

```diff
@@ -1,161 +1,213 @@
+	topic := constructTopic(client.realm, subscribeBuilder.topic)
+	if client.registerCallback(topic, subscribeBuilder.eventName, subscribeBuilder.callback) {
+		logrus.Warnf("topic and event already registered. event : %s", subscribeBuilder.eventName)
+		return
+	}
 
-	_, _, err := client.syncProducer.SendMessage(&sarama.ProducerMessage{
-		Timestamp: time.Now().UTC(),
-		Value: &eventEntry{
-			Event: *event,
-		},
-		Topic: event.Topic,
-	})
+	config := client.subscribeConfig
+	config.Topic = subscribeBuilder.topic
+	reader := kafka.NewReader(config)
+	if err := reader.SetOffset(kafka.LastOffset); err != nil {
+		logrus.Error("unable to set offset : ", err)
+		return
+	}
 
-	return err
+	// listening a topic
+	// each topic would have his own goroutine
+	go func() {
+		defer func() {
+			_ = reader.Close()
+		}()
+		for {
+			consumerMessage, err := reader.ReadMessage(subscribeBuilder.ctx)
+			if err != nil {
+				logrus.Error("unable to consume topic from kafka : ", err)
+				return
+			}
+			go client.processMessage(consumerMessage)
+		}
+	}()
 }
 
-// PublishEventAsync publish to a Kafka topic asynchronously
-func (client *KafkaClient) PublishEventAsync(event *Event) {
-	event.Realm = client.realm
-	event.Time = time.Now().UTC()
+// registerCallback add callback to map with topic and eventName as a key
+func (client *KafkaClient) registerCallback(topic, eventName string, callback func(event *Event, err error)) (
+	isRegistered bool) {
+	client.Mutex.Lock()
+	if client.subscribeMap == nil {
+		client.subscribeMap = make(map[string]map[string]func(*Event, error))
+	}
 
-	client.asyncProducer.Input() <- &sarama.ProducerMessage{
-		Timestamp: time.Now().UTC(),
-		Value: &eventEntry{
-			Event: *event,
-		},
-		Topic: event.Topic,
+	if _, isExists := client.subscribeMap[topic][eventName]; isExists {
+		client.Mutex.Unlock()
+		return true
 	}
+
+	client.subscribeMap[topic][eventName] = callback
+	client.Mutex.Unlock()
+	return false
 }
 
-// configureKafka configure kafka
-func configureKafka(kafkaConfig *KafkaConfig) (config *sarama.Config) {
-	config = sarama.NewConfig()
-	if kafkaConfig.DialTimeout != 0 {
-		config.Net.DialTimeout = kafkaConfig.DialTimeout
-	}
-	if kafkaConfig.ReadTimeout != 0 {
-		config.Net.ReadTimeout = kafkaConfig.ReadTimeout
-	}
-	if kafkaConfig.WriteTimeout != 0 {
-		config.Net.WriteTimeout = kafkaConfig.WriteTimeout
-	}
-	if kafkaConfig.MetadataRetryMax != 0 {
-		config.Metadata.Retry.Max = kafkaConfig.MetadataRetryMax
+// processMessage process a message from kafka
+func (client *KafkaClient) processMessage(message kafka.Message) {
+	event := unmarshal(message)
+	client.runCallback(event, message)
+}
+
+// unmarshal unmarshal received message into event struct
+func unmarshal(message kafka.Message) *Event {
+	var event Event
+	err := json.Unmarshal(message.Value, &event)
+	if err != nil {
+		logrus.Error("unable to unmarshal message from consumer in kafka : ", err)
```

Comment:
why don't we just return error and let the caller decide what to do?

Reply:
yes, it will better if we return the error 
thanks